primitives/src/sync.rs

```diff
@@ -0,0 +1,149 @@
+use core::ops::{Deref, DerefMut};
+use std::{
+    sync::atomic::{AtomicU32, Ordering},
+    time::{Duration, Instant},
+};
+
+const TIME_SPINNING_MICRO_SECONDS: u128 = 500;
+const SLEEP_TIME: Duration = Duration::from_micros(100);
+
+pub struct MutexMutGuard<'a, T: ?Sized> {
+    mutex: &'a Mutex<T>,
+}
+
+pub struct MutexGuard<'a, T: ?Sized> {
+    mutex: &'a Mutex<T>,
+}
+
+pub struct Mutex<T: ?Sized> {
+    counter: AtomicU32,
+    interior: T,
+}
+
```

Comment:
This looks like you're implementing a custom `Mutex`. When in reality you should be implementing a thin wrapper for a pre-existing Mutex implementation.

Reply:
`parking_lot` offers [`Mutex`](https://docs.rs/parking_lot/latest/parking_lot/type.Mutex.html#) without poisoning.